From d2eaa7479fb5e5c3171c80f686048c166ee18fe1 Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Tue, 13 Mar 2018 21:09:12 +0000
Subject: [PATCH] moray backend and mapping work bumped ipaddr.js version to
 1.6.0 started portolan command changes serializer, some integration tests,
 client, mapping and more

---
 bin/portolan                     | 119 +++++++++++++++++++++++++++++++
 etc/svp-types.json               |  32 +++++++--
 lib/backend/json.js              |  37 ++++++++++
 lib/backend/moray.js             | 101 +++++++++++++++++++++++++-
 lib/client.js                    |  49 +++++++++++++
 lib/common.js                    |  58 +++++++++++++++
 lib/framer.js                    |   2 +
 lib/parser.js                    |  20 ++++++
 lib/serialize.js                 |  22 ++++++
 lib/types.js                     |  19 +++--
 package.json                     |   4 +-
 test/integration/backend.test.js | 109 ++++++++++++++++++++++++++++
 test/lib/mapping.js              |  34 ++++++++-
 test/lib/request.js              |  46 +++++++++++-
 test/unit/types.test.js          |  19 +++++
 15 files changed, 658 insertions(+), 13 deletions(-)

diff --git a/bin/portolan b/bin/portolan
index 4de1039..97f5106 100755
--- a/bin/portolan
+++ b/bin/portolan
@@ -57,6 +57,12 @@ var OPTS = {
         help: 'IP address'
     },
 
+    r_ip: {
+        names: [ 'r_ip' ],
+        type: 'string',
+        help: 'remote IP address'
+    },
+
     mac: {
         names: [ 'mac' ],
         type: 'string',
@@ -86,6 +92,26 @@ var OPTS = {
         type: 'arrayOfString',
         help: 'array of log ids to remove'
     }
+
+    /*
+    r_dc_id: {
+        names: [ 'r_dc_id' ],
+        type: 'positiveInteger',
+        help: 'remote datacenter id'
+    },
+
+    r_vnet_id: {
+        names: [ 'r_vnet_id' ],
+        type: 'positiveInteger',
+        help: 'remote virtual network ID'
+    },
+
+    r_vlan_id: {
+        names: [ 'r_vlan_id' ],
+        type: 'positiveInteger',
+        help: 'remote vlan ID'
+    }
+    */
 };
 
 
@@ -437,5 +463,98 @@ PortolanCLI.prototype['do_log_rm'].options = [
     OPTS.id
 ];
 
+/**
+ * Add an vnet route mapping
+ */
+PortolanCLI.prototype['do_add-vnetroute'] =
+    function _addVnetRoute(_subcmd, opts, _args, callback) {
+
+    var backendOpts;
+    try {
+        backendOpts = translateOpts(opts, {
+            vnet_id: 'vnet_id',
+            vlan_id: 'vlan_id',
+            subnet: 'subnet',
+            r_dc_id: 'r_dc_id',
+            r_vnet_id: 'r_vnet_id',
+            r_vlan_id: 'r_vlan_id',
+            r_subnet: 'r_subnet',
+            r_send_mac: 'r_send_mac'
+        });
+        mod_common.validate.opts(backendOpts);
+    } catch (optsErr) {
+        callback(optsErr);
+        return;
+    }
+
+    backendOpts.subnet = mod_common.IPv6SubObj(backendOpts.subnet);
+    backendOpts.r_subnet = mod_common.IPv6SubObj(backendOpts.r_subnet);
+    backendOpts.r_send_mac = mod_common.macToInt(backendOpts.r_send_mac);
+
+    mod_backend.addVnetRouteMapping(backendOpts, callback);
+};
+
+PortolanCLI.prototype['do_add-vnetroute'].options = [
+    OPTS.vnet_id,
+    OPTS.vlan_id,
+    OPTS.subnet,
+    OPTS.r_dc_id,
+    OPTS.r_vnet_id,
+    OPTS.r_vlan_id,
+    OPTS.r_subnet,
+    OPTS.r_send_mac
+];
+
+/**
+ * Do a L3 lookup
+PortolanCLI.prototype['do_vnetroute-lookup'] =
+    function _vnetRouteReq(_subcmd, opts, _args, callback) {
+
+    // TODO
+    return;
+    var connOpts;
+    try {
+        connOpts = translateOpts(opts, {
+            // TODO
+            ip: 'ip',
+            vnet_id: 'vnet_id'
+        });
+        mod_common.validate.opts(connOpts);
+    } catch (optsErr) {
+        callback(optsErr);
+        return;
+    }
+
+    setClientOpts(opts, connOpts);
+    connOpts.ip = mod_common.IPv6obj(opts.ip);
+
+    var client = mod_client.connect(connOpts, function _afterVl3Conn(err) {
+        if (err) {
+            callback(err);
+            return;
+        }
+
+        client.vl3Req(connOpts, function (err2, res) {
+            if (err2) {
+                callback(err2);
+                return;
+            }
+
+            json(res);
+            callback();
+        });
+    });
+};
+
+PortolanCLI.prototype['do_vnetroute-lookup'].options = [
+    OPTS.host,
+    OPTS.ip,
+    OPTS.port,
+    OPTS.vnet_id
+];
+ */
+
+// TODO: add vnetroute-remove
+
 var cli = new PortolanCLI();
 mod_cmdln.main(cli);
diff --git a/etc/svp-types.json b/etc/svp-types.json
index 7095730..be0d09b 100644
--- a/etc/svp-types.json
+++ b/etc/svp-types.json
@@ -4,7 +4,14 @@
 		"created_at": 1430352621,
 		"derived_from": "src/types",
 		"ctf_version": 2,
-		"requested_types": [ "svp_req_t" , "svp_vl2_req_t" , "svp_vl2_ack_t" , "svp_vl3_req_t" , "svp_vl3_ack_t" , "svp_bulk_req_t" , "svp_bulk_ack_t" , "svp_log_req_t" , "svp_log_vl2_t" , "svp_log_vl3_t" , "svp_log_ack_t" , "svp_lrm_req_t" , "svp_lrm_ack_t" , "svp_shootdown_t" ]
+		"requested_types": [
+            "svp_req_t", "svp_vl2_req_t", "svp_vl2_ack_t",
+            "svp_vl3_req_t", "svp_vl3_ack_t", "svp_bulk_req_t",
+            "svp_bulk_ack_t", "svp_log_req_t", "svp_log_vl2_t",
+            "svp_log_vl3_t", "svp_log_ack_t", "svp_lrm_req_t",
+            "svp_lrm_ack_t", "svp_shootdown_t", "svp_route_req_t",
+            "svp_route_ack_t"
+        ]
 	},
 "data":
 	[
@@ -55,8 +62,7 @@
 			{ "name": "sl3a_uip", "type": "uint8_t [16]" }
 		] },
 		{ "name": "svp_vl3_req_t", "typedef": "struct svp_vl3_req" },
-		{ "name": "svp_vl3_ack_t", "typedef": "struct svp_vl3_ack" },
-		{ "name": "struct svp_bulk_req", "struct": [
+		{ "name": "svp_vl3_ack_t", "typedef": "struct svp_vl3_ack" }, { "name": "struct svp_bulk_req", "struct": [
 			{ "name": "svbr_type", "type": "uint32_t" }
 		] },
 		{ "name": "svp_bulk_req_t", "typedef": "struct svp_bulk_req" },
@@ -95,6 +101,24 @@
 			{ "name": "svl3_vnetid", "type": "uint32_t" }
 		] },
 		{ "name": "svp_log_vl3_t", "typedef": "struct svp_log_vl3" },
-		{ "name": "svp_shootdown_t", "typedef": "struct svp_shootdown" }
+		{ "name": "svp_shootdown_t", "typedef": "struct svp_shootdown" },
+		{ "name": "struct svp_route_req", "struct": [
+			{ "name": "srr_vnetid", "type": "uint32_t" },
+			{ "name": "srr_vlanid", "type": "uint16_t" },
+			{ "name": "srr_pad", "type": "uint16_t" },
+			{ "name": "srr_srcip", "type": "uint32_t [4]" },
+			{ "name": "srr_dstip", "type": "uint32_t [4]" }
+		] },
+		{ "name": "struct svp_route_ack", "struct": [
+			{ "name": "sra_status", "type": "uint32_t" },
+			{ "name": "sra_vnetid", "type": "uint32_t" },
+			{ "name": "sra_vlanid", "type": "uint16_t" },
+			{ "name": "sra_prefixlen", "type": "uint8_t" },
+			{ "name": "sra_pad", "type": "uint8_t" },
+			{ "name": "sra_dcid", "type": "uint32_t" },
+			{ "name": "sra_vl2_srcmac", "type": "uint8_t [6]" }
+		] },
+		{ "name": "svp_route_req_t", "typedef": "struct svp_route_req" },
+		{ "name": "svp_route_ack_t", "typedef": "struct svp_route_ack" }
 	]
 }
diff --git a/lib/backend/json.js b/lib/backend/json.js
index baaf92e..ff2cf54 100644
--- a/lib/backend/json.js
+++ b/lib/backend/json.js
@@ -58,6 +58,10 @@ var UNDERLAY_FILE = {
 };
 
 
+var ROUTE_FILE = {
+    // TODO
+};
+
 
 // --- Internal
 
@@ -177,6 +181,8 @@ JsonStore.prototype._transform = function _dssTransform(msg, _enc, callback) {
         return this.logReq(msg, callback);
     case types.svp_op.SVP_R_LOG_RM:
         return this.logRm(msg, callback);
+    case types.svp_op.SVP_R_ROUTE_REQ:
+        return this.routeReq(msg, callback);
     default:
         this.log.warn({ message: msg }, 'unsupported svp_type');
         // XXX: push some sort of error on here?
@@ -334,6 +340,37 @@ JsonStore.prototype.logRm = function _jsonLogRm(_msg, callback) {
     callback();
 };
 
+JsonStore.prototype.routeReq = function _jsonRouteReq(msg, callback) {
+    var self = this;
+
+    // TODO
+    return;
+
+    loadFile(ROUTE_FILE, function _afterRouteLoad(err, table) {
+        if (err) {
+            // XXX: what to do here?
+            callback();
+            return;
+        }
+        // TODO: some stuff
+
+        self.push({
+            svp_type: types.svp_op.SVP_R_ROUTE_ACK,
+            svp_id: msg.svp_id,
+            svp_msg: {
+                vnetid: null, // remote vnet id
+                vlan: null, // remote vlan id
+                prefixlen: null, // prefix of the remote vl3 subnet
+                pad: 0, //hummm
+                dcid: null, // remote datacenter id
+                vl2_srcmac: null //our vl2 source mac 
+            }
+        });
+        callback();
+        return;
+    });
+};
+
 // --- Exports
 
 
diff --git a/lib/backend/moray.js b/lib/backend/moray.js
index aed1a99..5163d8f 100644
--- a/lib/backend/moray.js
+++ b/lib/backend/moray.js
@@ -359,6 +359,64 @@ MorayStore.prototype._op_svp_r_log_rm =
 };
 
 
+MorayStore.prototype._op_svp_r_route_req =
+    function _op_svp_r_route_req(input, log, callback) {
+    var self = this;
+
+    mod_portolan_moray.vnetRouteLookup({
+        log: log,
+        moray: self.moray,
+        src_vnet_id: input.svp_msg.vnet_id,
+        src_vlan_id: input.svp_msg.vlan_id
+    }, function (vrErr, routes) {
+        if (vrErr) {
+            // TODO: add logging
+            callback(vrErr);
+            return;
+        }
+
+        var r_ip = input.svp_msg.r_ip;
+        var ip = input.svp_msg.ip;
+
+        for (var r in routes) {
+            if (!common.cidrContainsIP(r.subnet, ip) ||
+                !common.cidrContainsIp(r.r_subnet, r_ip)) {
+
+                continue;
+            }
+
+            /*
+             * Subnets are stored in IPv4 Mapped IPv6 format, so prefixlen
+             * should be > 96.
+             */
+            var split = r.r_subnet.split('/');
+            var prefixlen = parseInt(split('/')[1], 10);
+            if (split[0].indexOf('::ffff') === 0 && prefixlen >= 96) {
+                prefixlen -= 96;
+            }
+
+            var output = {
+                svp_msg: {
+                    sra_status: types.svp_status.SVP_S_OK,
+                    sra_vnetid: r.r_vnet_id,
+                    sra_vlanid: r.r_vlan_id,
+                    sra_prefixlen: prefixlen,
+                    sra_dcid: r.r_dc_id,
+                    sra_vl2_src_mac: r.r_send_mac
+                },
+                svp_type: types.svp_op.SVP_R_ROUTE_ACK
+            };
+
+            callback(null, output);
+            return;
+        }
+
+        callback();
+        return;
+    });
+};
+
+
 // --- Backend manipulation
 
 
@@ -435,6 +493,43 @@ function removeUnderlayMapping(opts, callback) {
     return mod_portolan_moray.removeUnderlayMapping(newOpts, callback);
 }
 
+function addVnetRouteMapping(opts, callback) {
+    assert.object(opts, 'opts');
+    assert.func(callback, 'callback');
+
+    var newOpts = clone(opts);
+    newOpts.subnet = common.cidrToString(opts.subnet);
+    newOpts.r_subnet = common.cidrToString(opts.r_subnet);
+    newOpts.moray = SHARED.moray;
+
+    return mod_portolan_moray.addVnetRouteMapping(newOpts, callback);
+}
+
+
+function updateVnetRouteMapping(opts, callback) {
+    assert.object(opts, 'opts');
+    assert.func(callback, 'callback');
+
+    var newOpts = clone(opts);
+    newOpts.subnet = common.cidrToString(opts.subnet);
+    newOpts.r_subnet = common.cidrToString(opts.r_subnet);
+    newOpts.moray = SHARED.moray;
+
+    return mod_portolan_moray.updateVnetRouteMapping(newOpts, callback);
+}
+
+
+function removeVnetRouteMapping(opts, callback) {
+    assert.object(opts, 'opts');
+    assert.func(callback, 'callback');
+
+    var newOpts = clone(opts);
+    newOpts.subnet = common.cidrToString(opts.subnet);
+    newOpts.r_subnet = common.cidrToString(opts.r_subnet);
+    newOpts.moray = SHARED.moray;
+
+    return mod_portolan_moray.removeVnetRouteMapping(newOpts, callback);
+}
 
 
 module.exports = {
@@ -448,5 +543,9 @@ module.exports = {
 
     addUnderlayMapping: addUnderlayMapping,
     updateUnderlayMapping: updateUnderlayMapping,
-    removeUnderlayMapping: removeUnderlayMapping
+    removeUnderlayMapping: removeUnderlayMapping,
+
+    addVnetRouteMapping: addVnetRouteMapping,
+    updateVnetRouteMapping: updateVnetRouteMapping,
+    removeVnetRouteMapping: removeVnetRouteMapping
 };
diff --git a/lib/client.js b/lib/client.js
index c4124c5..b55fae8 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -39,6 +39,22 @@ function PortolanClient(opts) {
 }
 
 
+/*
+ * This function, passes a request in the form of a JS object through the
+ * serializer, then to portolan as binary data, and parsers the response.  See
+ * the flow below:
+ *
+ * client: serialize request
+ * client: write request to portolan
+ *      portolan: Framer parses header
+ *      portolan: Parser parses payload
+ *      portolan: backend conducts get/set/update with moray or other backend
+ *      portolan: Serializes response
+ *      portolan: sends response
+ * client: Framer parses header
+ * client: Parser parses payload
+ * client: once('data', callback());
+ */
 PortolanClient.prototype._makeReq = function _makeReq(req, callback) {
     if (!this._serStream) {
         var strOpts = { log: this.log };
@@ -218,6 +234,39 @@ PortolanClient.prototype.logRm = function _clientLogRm(opts, callback) {
     });
 };
 
+
+PortolanClient.prototype.vnetRouteReq = function _clientVL3req(opts, callback) {
+    mod_assert.object(opts, 'opts');
+    mod_assert.number(opts.vnet_id, 'opts.vnet_id');
+    mod_assert.number(opts.vlan_id, 'opts.vlan_id');
+    mod_assert.object(opts.subnet, 'opts.subnet');
+    mod_assert.object(opts.r_subnet, 'opts.r_subnet');
+
+    this._makeReq({
+        svp_type: mod_types.svp_op.SVP_R_ROUTE_REQ,
+        svp_id: this.id++,
+        svp_msg: {
+            srr_vnetid: opts.vnet_id,
+            srr_vlanid: opts.vlan_id,
+            srr_srcip: opts.ip,
+            srr_dstip: opts.r_ip
+        }
+    }, function (data) {
+        var res = {
+            status: data.svp_msg.r_status,
+            status_str: mod_types.statusString(data.svp_msg.r_status),
+            r_vnet_id: data.svp_msg.r_vnet_id,
+            r_vlan_id: data.svp_msg.r_vlan_id,
+            r_prefixlen: data.svp_msg.r_prefixlen,
+            r_dc_id: data.svp_msg.r_dc_id,
+            r_send_mac: mod_common.intToMac(data.svp_msg.r_send_mac)
+        };
+
+        return callback(null, res);
+    });
+};
+
+
 // --- Exports
 
 
diff --git a/lib/common.js b/lib/common.js
index 677f9e9..22c0f07 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -14,6 +14,7 @@
 
 'use strict';
 
+// XXX: Switch to node-ip6addr
 var ipaddr = require('ipaddr.js');
 var types = require('./types');
 var fmt = require('util').format;
@@ -78,6 +79,33 @@ function createv6obj(ipStr) {
 }
 
 
+function createv6subobj(cidr) {
+    if (typeof (cidr) === 'object') {
+        return cidr;
+    }
+
+    var split = cidr.split('/');
+    var ip = split[0];
+    var prefixlen = parseInt(split[1], 10);
+    var ipStr;
+
+    var ipObj = ipaddr.process(ip);
+    if (ipObj.kind() === 'ipv4') {
+        if (prefixlen < 33) {
+            prefixlen += 96;
+        }
+        ipStr = ipObj.toIPv4MappedAddress().toString();
+    }
+
+    return ipaddr.parseCIDR(ipStr + '/' + prefixlen.toString());
+}
+
+
+function cidrToString(cidr) {
+    return (cidr.toString().replace(',', '/'));
+}
+
+
 function ipToString(ip) {
     return (ip.toString());
 }
@@ -95,6 +123,32 @@ function stringToIp(ip) {
     return (ipaddr.parse(ip));
 }
 
+
+function cidrContainsIp(cidr, ip) {
+    var addr = stringToIp(ip);
+    var split = cidr.split('/');
+    var prefixlen;
+
+    if (split.length !== 2) {
+        return false;
+    }
+
+    // XXX: what about a hex prefix len?
+    prefixlen = parseInt(split[1], 10);
+
+    /*
+     * Be flexible for IPv4 Mapped subnets with IPv4 prefix lengths
+     * This is a bit of a workaround for the short comings of ipaddr.js.
+     */
+    if (cidr.indexOf('::ffff') === 0 && prefixlen < 33) {
+        prefixlen += 96;
+        split[1] = prefixlen;
+        cidr = split.join('/');
+    }
+
+    return addr.match(ipaddr.parseCIDR(cidr));
+}
+
 /**
  * Convert an array of bytes to an integer
  */
@@ -236,6 +290,7 @@ function validateOpts(opts) {
     if (opts.hasOwnProperty('ip')) {
         validateIP(opts.ip);
     }
+    // TODO add subnet, vnet, vlan, dc_id
 }
 
 /**
@@ -317,9 +372,12 @@ module.exports = {
     childLogger: createChildLogger,
     fatalResponse: fatalResponse,
     IPv6obj: createv6obj,
+    IPv6SubObj: createv6subobj,
     ipToString: ipToString,
+    cidrToString: cidrToString,
     ipv4StrTov6: ipv4StrTov6,
     stringToIp: stringToIp,
+    cidrContainsIp: cidrContainsIp,
     macArrToInt: macArrToInt,
     macToArr: macToArr,
     macToInt: macToInt,
diff --git a/lib/framer.js b/lib/framer.js
index db04a5c..0d69f7a 100644
--- a/lib/framer.js
+++ b/lib/framer.js
@@ -96,6 +96,8 @@ SVPFramer.prototype._transform =
         log.debug({ msg_svp_req_hdr: msg.svp_req.hdr, msgCount: msgCount++ },
             'message header');
 
+        // XXX: version check? msg.svp_req.hdr.svp_ver
+
         // There is a distinction between what the message indicates as its
         // size, and the expected size of the message derived from ctypes.
         // We check the former here & the latter in the parser.
diff --git a/lib/parser.js b/lib/parser.js
index 65f5f4d..288c614 100644
--- a/lib/parser.js
+++ b/lib/parser.js
@@ -253,6 +253,26 @@ SVPparser.prototype._transform =
         };
         break;
 
+    case mod_types.svp_op.SVP_R_ROUTE_REQ:
+        rec.svp_msg = {
+            vnet_id: payload.body.srr_vnetid,
+            vlan_id: payload.body.srr_vlanid,
+            ip: payload.body.srr_srcip,
+            r_ip: payload.body.srr_dstip
+        };
+        break;
+
+    case mod_types.svp_op.SVP_R_ROUTE_ACK:
+        rec.svp_msg = {
+            r_status: payload.body.sra_status,
+            r_vnet_id: payload.body.sra_vlanid,
+            r_vlan_id: payload.body.sra_vlanid,
+            r_prefixlen: payload.body.sra_prefixlen,
+            r_dc_id: payload.body.sra_dcid,
+            r_send_mac: mod_common.macArrToInt(payload.body.sra_vl2_srcmac)
+        };
+        break;
+
     default:
         log.warn({ hdr: msg.svp_req.hdr, body: payload.body },
             'unknown message svp_op');
diff --git a/lib/serialize.js b/lib/serialize.js
index c990e40..a98402b 100644
--- a/lib/serialize.js
+++ b/lib/serialize.js
@@ -196,6 +196,28 @@ SVPserializer.prototype._transform =
         ];
         break;
 
+    case types.svp_op.SVP_R_ROUTE_ACK:
+        fixedPart = [
+            msg.svp_msg.sra_status,
+            msg.svp_msg.sra_vnetid,
+            msg.svp_msg.sra_vlanid,
+            msg.svp_msg.sra_prefixlen,
+            [ 0 ], // padding
+            msg.svp_msg.sra_dcid,
+            common.intToMacArrayOfNums(msg.svp_msg.sra_vl2_src_mac)
+        ];
+        break;
+
+    case types.svp_op.SVP_R_ROUTE_REQ:
+        fixedPart = [
+            msg.svp_msg.srr_vnetid,
+            msg.svp_msg.srr_vlanid,
+            [ 0, 0 ], // padding
+            msg.svp_msg.srr_srcip.toByteArray(),
+            msg.svp_msg.srr_dstip.toByteArray()
+        ];
+        break;
+
     default:
         log.warn({ message: msg }, 'unknown message svp_type');
         return callback();
diff --git a/lib/types.js b/lib/types.js
index 4733df1..2fef50a 100644
--- a/lib/types.js
+++ b/lib/types.js
@@ -31,7 +31,10 @@ var op_types = {
     9: 'svp_log_req_t',
     10: 'svp_log_ack_t',
     11: 'svp_lrm_req_t',
-    12: 'svp_lrm_ack_t'
+    12: 'svp_lrm_ack_t',
+    13: 'svp_shootdown_t',
+    14: 'svp_route_req_t',
+    15: 'svp_route_ack_t'
 };
 
 var sizeof = {
@@ -48,7 +51,10 @@ var sizeof = {
     SVP_LOG_VL2: 32,
     SVP_LOG_VL3: 44,
     SVP_R_LOG_RM: 4,
-    SVP_R_LOG_RM_ACK: 4
+    SVP_R_LOG_RM_ACK: 4,
+    SVP_R_SHOOTDOWN: 12,
+    SVP_R_ROUTE_REQ: 40,
+    SVP_R_ROUTE_ACK: 18
 };
 
 var sizeof_ops = {
@@ -61,7 +67,10 @@ var sizeof_ops = {
     9: sizeof.SVP_R_LOG_REQ,
     10: sizeof.SVP_R_LOG_ACK,
     11: sizeof.SVP_R_LOG_RM,
-    12: sizeof.SVP_R_LOG_RM_ACK
+    12: sizeof.SVP_R_LOG_RM_ACK,
+    13: sizeof.SVP_R_SHOOTDOWN,
+    14: sizeof.SVP_R_ROUTE_REQ,
+    15: sizeof.SVP_R_ROUTE_ACK
 };
 
 var svp_op = {
@@ -88,7 +97,9 @@ var svp_op_hex = {
     SVP_R_LOG_ACK: '0x0A',
     SVP_R_LOG_RM: '0x0B',
     SVP_R_LOG_RM_ACK: '0x0C',
-    SVP_R_SHOOTDOWN: '0x0D'
+    SVP_R_SHOOTDOWN: '0x0D',
+    SVP_R_ROUTE_REQ: '0x0E',
+    SVP_R_ROUTE_ACK: '0x0F'
 };
 
 var svp_status = {
diff --git a/package.json b/package.json
index 219721f..96fc2ac 100644
--- a/package.json
+++ b/package.json
@@ -11,13 +11,13 @@
         "cmdln": "3.0.1",
         "ctype": "0.5.3",
         "dtrace-provider": "0.3.1",
-        "ipaddr.js": "0.1.6",
+        "ipaddr.js": "1.6.0",
         "lomstream": "^1.1.0",
         "lru-cache": "2.5.0",
         "mkdirp": "0.5.0",
         "moray": "3.4.2",
         "node-uuid": "1.4.2",
-        "portolan-moray": "git+https://github.com/joyent/sdc-portolan-moray.git#7e2c4ca",
+        "portolan-moray": "git+https://github.com/joyent/sdc-portolan-moray.git#vpc",
         "tape": "3.0.3",
         "vasync": "1.6.2",
         "verror": "1.6.0",
diff --git a/test/integration/backend.test.js b/test/integration/backend.test.js
index 7ea3c81..efbb984 100644
--- a/test/integration/backend.test.js
+++ b/test/integration/backend.test.js
@@ -51,6 +51,40 @@ var VMS = [
     }
 ];
 
+var ROUTES = [
+    {
+        vnet_id: 11111,
+        vlan_id: 1,
+        subnet: '192.168.111.0/24',
+        r_dc_id: 22,
+        r_vnet_id: 22222,
+        r_vlan_id: 2,
+        r_subnet: '192.168.222.0/24',
+        r_send_mac: '00:0a:95:ff:ff:ff'
+    },
+    {
+        vnet_id: 101010,
+        vlan_id: 10,
+        subnet: '192.168.100.0/24',
+        r_dc_id: 220,
+        r_vnet_id: 202020,
+        r_vlan_id: 20,
+        r_subnet: '192.168.200.0/24',
+        r_send_mac: '00:0a:59:ff:ff:ff'
+    }
+];
+
+var ROUTE_IPS = [
+    {
+        src: '192.168.111.1',
+        dst: '192.168.222.1'
+    },
+    {
+        src: '192.168.100.1',
+        dst: '192.168.200.1'
+    }
+];
+
 var STATUS = mod_types.svp_status;
 
 
@@ -80,6 +114,18 @@ test('setup', function (t) {
             params: VMS[1]
         });
     });
+
+    t.test('add vnet route mapping 0', function (t2) {
+        mod_mapping.addVnetRoute(t2, {
+            params: ROUTES[0]
+        });
+    });
+
+    t.test('add vnet route mapping 1', function (t2) {
+        mod_mapping.addVnetRoute(t2, {
+            params: ROUTES[1]
+        });
+    });
 });
 
 
@@ -193,6 +239,69 @@ test('vl3', function (t) {
 
 });
 
+test('vnetRoute', function (t) {
+    t.test('mapping exists', function (t2) {
+        mod_req.vnetRoute(t2, {
+            params: {
+                vnet_id: ROUTES[0].vnet_id,
+                vlan_id: ROUTES[0].vlan_id,
+                subnet: ROUTE_IPS[0].src,
+                r_subnet: ROUTE_IPS[0].dst
+            },
+            exp: {
+                status: STATUS.SVP_S_OK,
+                status_str: mod_types.statusString(STATUS.SVP_S_OK),
+                vnet_id: ROUTES[0].vnet_id,
+                vlan_id: ROUTES[0].vlan_id,
+                subnet: ROUTES[0].subnet,
+                r_dc_id: ROUTES[0].r_dc_id,
+                r_vnet_id: ROUTES[0].r_vnet_id,
+                r_vlan_id: ROUTES[0].r_vlan_id,
+                r_subnet: ROUTES[0].r_subnet,
+                r_send_mac: ROUTES[0].r_send_mac
+            }
+        });
+    });
+
+
+    t.test('wrong source ip/subnet', function (t2) {
+        mod_req.vnetRoute(t2, {
+            params: {
+                vnet_id: ROUTES[0].vnet_id,
+                vlan_id: ROUTES[0].vlan_id,
+                ip: '10.10.10.10',
+                r_ip: ROUTE_IPS[0].dst
+            },
+            exp: mod_req.vnetRouteNotFound()
+        });
+    });
+
+// TODO others
+
+    /*
+    t.test('IP exists, but not vnet_id', function (t2) {
+        mod_req.vl3(t2, {
+            params: {
+                ip: VMS[0].ip,
+                vnet_id: VMS[0].vnet_id + 1
+            },
+            exp: mod_req.vl3NotFound()
+        });
+    });
+
+
+    t.test('overlay mapping exists, but not underlay', function (t2) {
+        mod_req.vl3(t2, {
+            params: {
+                ip: VMS[1].ip,
+                vnet_id: VMS[1].vnet_id
+            },
+            exp: mod_req.vl3NotFound()
+        });
+    });
+    */
+
+});
 
 // XXX: log lookup tests:
 // - try to delete a record that doesn't exist
diff --git a/test/lib/mapping.js b/test/lib/mapping.js
index 62b37ed..1b002a7 100644
--- a/test/lib/mapping.js
+++ b/test/lib/mapping.js
@@ -65,9 +65,41 @@ function addUnderlayMapping(t, opts) {
     });
 }
 
+function addVnetRouteMapping(t, opts) {
+    assert.object(t, 't');
+    assert.object(opts, 'opts');
+    assert.object(opts.params, 'opts.params');
+    assert.number(opts.params.vnet_id, 'opts.params.vnet_id');
+    assert.number(opts.params.vlan_id, 'opts.params.vlan_id');
+    assert.string(opts.params.subnet, 'opts.params.subnet');
+    assert.number(opts.params.r_dc_id, 'opts.params.r_dc_id');
+    assert.number(opts.params.r_vnet_id, 'opts.params.r_vnet_id');
+    assert.number(opts.params.r_vlan_id, 'opts.params.r_vlan_id');
+    assert.string(opts.params.r_subnet, 'opts.params.r_subnet');
+    assert.string(opts.params.r_send_mac, 'opts.params.r_send_mac');
+
+    var val = {
+        vnet_id: opts.params.vnet_id,
+        vlan_id: opts.params.vlan_id,
+        subnet: mod_common.IPv6SubObj(opts.params.subnet),
+        r_dc_id: opts.params.r_dc_id,
+        r_vnet_id: opts.params.r_vnet_id,
+        r_vlan_id: opts.params.r_vlan_id,
+        r_subnet: mod_common.IPv6SubObj(opts.params.r_subnet),
+        r_send_mac: mod_common.macToInt(opts.params.r_send_mac)
+    };
+
+    mod_moray.addVnetRouteMapping(val, function _afterVnetRoute(err) {
+        t.ifErr(err, 'add vnet route mapping');
+        t.end();
+        return;
+    });
+}
+
 
 
 module.exports = {
     addOverlay: addOverlayMapping,
-    addUnderlay: addUnderlayMapping
+    addUnderlay: addUnderlayMapping,
+    addVnetRoute: addVnetRouteMapping
 };
diff --git a/test/lib/request.js b/test/lib/request.js
index c69ad92..90ce258 100644
--- a/test/lib/request.js
+++ b/test/lib/request.js
@@ -97,6 +97,34 @@ function reqVL3(t, opts) {
     });
 }
 
+/**
+ * Make a vnetroute request
+ */
+function reqVnetRoute(t, opts) {
+    assert.object(t, 't');
+    assert.object(opts, 'opts');
+    assert.object(opts.exp, 'opts.exp');
+    assert.object(opts.params, 'opts.params');
+
+    mod_client.get(function (_, client) {
+        var params = {
+            vnet_id: opts.params.vnet_id,
+            vlan_id: opts.params.vlan_id,
+            ip: mod_common.IPv6obj(opts.params.ip),
+            r_ip: mod_common.IPv6obj(opts.params.r_ip)
+        };
+
+        client.vnetRouteReq(params, function _afterVL3(err, res) {
+            t.ifErr(err, 'vnetRoute error');
+            if (err) {
+                return t.end();
+            }
+
+            t.deepEqual(res, opts.exp, 'vnetRoute res');
+            return t.end();
+        });
+    });
+}
 
 /**
  * Returns a client response for VL2 mapping not found
@@ -124,6 +152,20 @@ function vl3NotFound() {
     };
 }
 
+function vnetRouteNotFound() {
+    return {
+        status: STATUS.SVP_S_NOTFOUND,
+        status_str: mod_types.statusString(STATUS.SVP_S_NOTFOUND),
+        vnet_id: 0,
+        vlan_id: 0,
+        subnet: '::',
+        r_dc_id: 0,
+        r_vnet_id: 0,
+        r_vlan_id: 0,
+        r_subnet: '::',
+        r_send_mac: '00:00:00:00:00:00'
+    };
+}
 
 
 module.exports = {
@@ -131,5 +173,7 @@ module.exports = {
     vl2: reqVL2,
     vl2NotFound: vl2NotFound,
     vl3: reqVL3,
-    vl3NotFound: vl3NotFound
+    vl3NotFound: vl3NotFound,
+    vnetRoute: reqVnetRoute,
+    vnetRouteNotFound: vnetRouteNotFound
 };
diff --git a/test/unit/types.test.js b/test/unit/types.test.js
index 315ae21..aa6e879 100644
--- a/test/unit/types.test.js
+++ b/test/unit/types.test.js
@@ -89,6 +89,24 @@ test('opinfo', function (t) {
             op: svp_op.SVP_R_LOG_RM_ACK,
             sizeof: 4,
             type: 'svp_lrm_ack_t'
+        },
+        {
+            name: 'SVP_R_SHOOTDOWN',
+            op: svp_op.SVP_R_SHOOTDOWN,
+            sizeof: 12,
+            type: 'svp_shootdown_t'
+        },
+        {
+            name: 'SVP_R_ROUTE_REQ',
+            op: svp_op.SVP_R_ROUTE_REQ,
+            sizeof: 40,
+            type: 'svp_route_req_t'
+        },
+        {
+            name: 'SVP_R_ROUTE_ACK',
+            op: svp_op.SVP_R_ROUTE_ACK,
+            sizeof: 18,
+            type: 'svp_route_ack_t'
         }
     ];
 
@@ -99,6 +117,7 @@ test('opinfo', function (t) {
         t.equal(Object.keys(type).length, 4, 'keys');
 
         var opinfo = mod_types.opInfo(type.op);
+
         t.equal(opinfo.name, type.name, type.name + ' name');
         t.equal(opinfo.sizeof, type.sizeof, 'sizeof');
         t.equal(opinfo.sizeofReq, type.sizeof + req_size, 'sizeofReq');
-- 
2.21.0

